bundle/deploy/state_pull.go

```diff
@@ -0,0 +1,146 @@
+package deploy
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+
+	"github.com/databricks/cli/bundle"
+	"github.com/databricks/cli/libs/filer"
+	"github.com/databricks/cli/libs/log"
+	"github.com/databricks/cli/libs/sync"
+)
+
+type statePull struct {
+	filerFactory
+}
+
+func (s *statePull) Apply(ctx context.Context, b *bundle.Bundle) error {
+	f, err := s.filerFactory(b)
+	if err != nil {
+		return err
+	}
+
+	// Download deployment state file from filer to local cache directory.
+	log.Infof(ctx, "Opening remote deployment state file")
+	remote, err := s.remoteState(ctx, f)
+	if err != nil {
+		log.Infof(ctx, "Unable to open remote deployment state file: %s", err)
+		return err
+	}
+	if remote == nil {
+		log.Infof(ctx, "Remote deployment state file does not exist")
+		return nil
+	}
+
+	statePath, err := getPathToStateFile(ctx, b)
+	if err != nil {
+		return err
+	}
+
+	local, err := os.OpenFile(statePath, os.O_CREATE|os.O_RDWR, 0600)
+	if err != nil {
+		return err
+	}
+	defer local.Close()
+
+	data := remote.Bytes()
+	if !isLocalStateStale(local, bytes.NewReader(data)) {
+		log.Infof(ctx, "Local deployment state is the same or newer, ignoring remote state")
+		return nil
+	}
+
+	// Truncating the file before writing
+	local.Truncate(0)
+	local.Seek(0, 0)
+
+	// Write file to disk.
+	log.Infof(ctx, "Writing remote deployment state file to local cache directory")
+	_, err = io.Copy(local, bytes.NewReader(data))
+	if err != nil {
+		return err
+	}
+
+	cacheDir, err := b.CacheDir(ctx)
+	if err != nil {
+		return fmt.Errorf("cannot get bundle cache directory: %w", err)
+	}
+
+	opts := &sync.SyncOptions{
+		LocalPath:        b.Config.Path,
+		RemotePath:       b.Config.Workspace.FilePath,
+		SnapshotBasePath: cacheDir,
+		Host:             b.WorkspaceClient().Config.Host,
+	}
+
+	snapshotPath, err := sync.SnapshotPath(opts)
+	if err != nil {
+		return err
+	}
+
+	if _, err := os.Stat(snapshotPath); err == nil {
```

Comment:
Shouldn't we always overwrite if the remote deployment state is newer? Otherwise we may not delete files that have been synchronized by another process.